src/Sarif.WorkItems/SarifWorkItemFiler.cs

```diff
@@ -0,0 +1,206 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Sarif.Visitors;
+using Microsoft.TeamFoundation.WorkItemTracking.WebApi.Models;
+using Microsoft.VisualStudio.Services.CircuitBreaker;
+using Microsoft.WorkItems;
+using Newtonsoft.Json;
+
+namespace Microsoft.CodeAnalysis.Sarif.WorkItems
+{
+    public class SarifWorkItemFiler
+    {
+        public SarifWorkItemFiler() { }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SarifWorkItemFiler"> class.</see>
+        /// </summary>
+        /// <param name="filingClient">
+        /// A client for communicating with a work item filing host (for example, GitHub or Azure DevOps).
+        /// </param>
+        /// <param name="filingContext">
+        /// A root context object that configures the work item filing operation.
+        /// </param>
+        public SarifWorkItemFiler(SarifWorkItemContext filingContext)
+        {
+            this.FilingContext = filingContext ?? throw new ArgumentNullException(nameof(filingContext));
+            this.FilingClient = FilingClientFactory.CreateFilingTarget(filingContext.ProjectUri);
+        }
+
+        public SarifWorkItemContext FilingContext { get; set; }
+
+        public FilingClient FilingClient { get; set; }
+
+        public virtual void FileWorkItems(Uri sarifLogFileLocation)
+        {
+            sarifLogFileLocation = sarifLogFileLocation ?? throw new ArgumentNullException(nameof(sarifLogFileLocation));
+
+            if (sarifLogFileLocation.IsAbsoluteUri && sarifLogFileLocation.Scheme == "file:")
+            {
+                FileWorkItems(File.ReadAllText(sarifLogFileLocation.LocalPath));
```

Comment:
>File.ReadAllText(sarifLogFileLocation.LocalPath) [](start = 30, length = 48)

As this set of overloads stands, you end up with both the string representation and the object model representation of the log in memory at the same time. To reduce memory usage, you can deserialize directly from the file stream into the object model:

```
if (sarifLogFileLocation.IsAbsoluteUri && sarifLogFileLocation.Scheme == "file:")
{
    using (var stream = new FileStream(sarifLogFileLocation.LocalPath, FileMode.Open, FileAccess.Read))
    using (var reader = new StreamReader(stream))
    using (var jsonReader = new JsonTextReader(reader))
    {
        var serializer = new JsonSerializer();
        SarifLog sarifLog = serializer.Deserialize<SarifLog>(jsonReader);
        FileWorkItems(sarifLog);
        }
}
```

Then perhaps you don't need the `FileWorkItems(string)` overload at all.

Reply:
Took the change. These overloads are all just sketched out for now.

---
In reply to: [381467520](https://github.com/microsoft/sarif-sdk/pull/1775#discussion_r381467520) [](ancestors = 381467520)